enzyme/Enzyme/GradientUtils.h

```diff
@@ -1531,6 +1531,217 @@ class GradientUtils : public CacheUtility {
         getNewFromOriginal(Builder2.getCurrentDebugLocation()));
     Builder2.setFastMathFlags(getFast());
   }
+
+  /// Returns the type for the internal representation of a derivative vector.
+  Type *getTypeForVectorMode(Type *ty) {
+    return getTypeForVectorMode(ty, width);
+  }
+
+  /// Returns the type for the internal representation of a derivative vector of
+  /// size `width`
+  static Type *getTypeForVectorMode(Type *ty, unsigned int width) {
+    if (ty->isIntegerTy() || ty->isFloatingPointTy()) {
```

Comment:
I'm not sure we want to actually use llvm vector types here as only certain widths are permitted per platform. Thus could we always use the arraytype?

Reply:
The comment is outdated. This does not concern the internal representation anymore, which is always an Array.  It is still used when we want to work with those unwrapped Arrays.
> only certain widths are permitted per platform
Isn't that code gens problem\?


On another note should we even vectorise add ourself or rely on llvm to vectors our n add instructions?

Reply:
Is this code even used then? If so can we remove it.

I would argue this is not code gen's problem yet since us creating an invalid vector (say 3) would not be something we can expect will be fixed. However upgrading to a vector from an array is something we should be able to rely on.

Reply:
So we should basically never emit any vectors and rely on the llvm vectoriser and instselct to do the right thing.

Reply:
For the moment that’s what I’d say. I think in the future we can enable using a vector if the width is available on the machine after doing a check.

Reply:
done

Reply:
I would still keep this function and use it everywhere (e.g no get width>1 array type otherwise regular.

However I might rename it get shadowtype